source/s3_client.c

```diff
@@ -1946,6 +1950,47 @@ static void s_s3_client_on_host_resolver_address_resolved(
     }
 }
 
+void aws_s3_client_aggregate_throughput(
+    struct aws_s3_client *client,
+    size_t throughput,
+    double *current_throughput_gbps,
+    double *running_throughput_average_gbps) {
+
+    struct aws_date_time current_timestamp;
+    aws_date_time_init_now(&current_timestamp);
+    uint64_t current_timestamp_millis = aws_date_time_as_millis(&current_timestamp);
+
+    aws_s3_client_lock_synced_data(client);
+    client->synced_data.pending_throughput_bytes += throughput;
+
+    double time_delta = (current_timestamp_millis - client->synced_data.throughput_timestamp_millis) / 1000.0;
+
+    if (time_delta > 1.0) {
+        client->synced_data.throughput_timestamp_millis = current_timestamp_millis;
+
+        double rate_in_bytes = (double)client->synced_data.pending_throughput_bytes / time_delta;
+        client->synced_data.current_throughput_gbps = rate_in_bytes * 8.0 / 1000.0 / 1000.0 / 1000.0;
+
+        if (client->synced_data.running_throughput_average_gbps == 0.0) {
```

Comment:
You probably don't actually need this check. It'll ramp up to the right place really fast.